src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -317,43 +370,81 @@ private void OnWriteCompleted(WriteContext writeContext)
+            }
+            else
             {
-                var waitingTask = _tasksPending.Dequeue();
-                var bytesToWrite = waitingTask.BytesToWrite;
+                _log.ConnectionWriteCallback(_connectionId, status);
+            }
+        }
+
+        private void CompleteNextWrite(ref int bytesLeftToBuffer)
+        {
+            // Called inside _contextLock
+            var waitingTask = _tasksPending.Dequeue();
+            var bytesToWrite = waitingTask.BytesToWrite;
 
-                _numBytesPreCompleted += bytesToWrite;
-                bytesLeftToBuffer -= bytesToWrite;
+            _numBytesPreCompleted += bytesToWrite;
+            bytesLeftToBuffer -= bytesToWrite;
 
-                if (_lastWriteError == null)
+            // Dispose registration if there is one
+            waitingTask.CancellationRegistration?.Dispose();
+
+            if (waitingTask.CancellationToken.IsCancellationRequested)
+            {
+                if (waitingTask.IsSync)
                 {
-                    if (waitingTask.IsSync)
-                    {
-                        waitingTask.CompletionSource.TrySetResult(null);
-                    }
-                    else
-                    {
-                        _threadPool.Complete(waitingTask.CompletionSource);
-                    }
+                    waitingTask.CompletionSource.TrySetCanceled();
                 }
                 else
                 {
-                    if (waitingTask.IsSync)
-                    {
-                        waitingTask.CompletionSource.TrySetException(_lastWriteError);
-                    }
-                    else
-                    {
-                        _threadPool.Error(waitingTask.CompletionSource, _lastWriteError);
-                    }
+                    _threadPool.Cancel(waitingTask.CompletionSource);
+                }
+            }
+            else
+            {
+                if (waitingTask.IsSync)
+                {
+                    waitingTask.CompletionSource.TrySetResult(null);
+                }
+                else
+                {
+                    _threadPool.Complete(waitingTask.CompletionSource);
                 }
             }
+        }
 
-            _log.ConnectionWriteCallback(_connectionId, status);
+        private void CompleteFinishedWrites(int status)
+        {
+            // Called inside _contextLock
+            // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
+            // This allows large writes to complete once they've actually finished.
+            var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
+            while (_tasksPending.Count > 0 &&
+                   (_tasksPending.Peek().BytesToWrite) <= bytesLeftToBuffer)
+            {
+                CompleteNextWrite(ref bytesLeftToBuffer);
+            }
+        }
+
+        private void CompleteAllWrites()
```

Comment:
It seems to me that this is only getting called if the user passes in a canceled token, but that's only one scenario where this is getting called.

We need to complete all writes whenever there is an error, even if we don't wind up cancelling them.